
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      <link rel="icon" href="../nesi_images/nesi_ga.png">
      <meta name="generator" content="mkdocs-1.3.0, mkdocs-material-8.3.9">
    
    
      
        <title>RNA-seq Mapping And Count Data Workflow - Introduction to Bash Scripting and HPC Scheduler</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.1d29e8d0.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.cbb835fc.min.css">
        
      
      
    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Mukta:300,300i,400,400i,700,700i%7CFira+Code:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Mukta";--md-code-font:"Fira Code"}</style>
      
    
    
    <script>__md_scope=new URL("..",location),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="teal" data-md-color-accent="">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#rna-seq-mapping-and-count-data-workflow" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Introduction to Bash Scripting and HPC Scheduler" class="md-header__button md-logo" aria-label="Introduction to Bash Scripting and HPC Scheduler" data-md-component="logo">
      
  <img src="../nesi_images/nesi_ga.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Introduction to Bash Scripting and HPC Scheduler
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              RNA-seq Mapping And Count Data Workflow
            
          </span>
        </div>
      </div>
    </div>
    
      <form class="md-header__option" data-md-component="palette">
        
          
          
          <input class="md-option" data-md-color-media="(prefers-color-scheme)" data-md-color-scheme="" data-md-color-primary="teal" data-md-color-accent=""  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
          
            <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_3" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m14.3 16-.7-2h-3.2l-.7 2H7.8L11 7h2l3.2 9h-1.9M20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69m-9.15 3.96h2.3L12 9l-1.15 3.65Z"/></svg>
            </label>
          
        
          
          
          <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent=""  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_2">
          
            <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
            </label>
          
        
          
          
          <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="teal" data-md-color-accent=""  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_3">
          
            <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_2" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
            </label>
          
        
      </form>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/GenomicsAotearoa/Workshop-Bash_Scripting_And_HPC_Job_Scheduler.git" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GenomicsAotearoa/Workshop-Bash_Scripting_And_HPC_Job_Scheduler
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Introduction to Bash Scripting and HPC Scheduler" class="md-nav__button md-logo" aria-label="Introduction to Bash Scripting and HPC Scheduler" data-md-component="logo">
      
  <img src="../nesi_images/nesi_ga.png" alt="logo">

    </a>
    Introduction to Bash Scripting and HPC Scheduler
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/GenomicsAotearoa/Workshop-Bash_Scripting_And_HPC_Job_Scheduler.git" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GenomicsAotearoa/Workshop-Bash_Scripting_And_HPC_Job_Scheduler
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        Introduction to Bash Scripting and HPC Scheduler
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../1_DesigningVariantC/" class="md-nav__link">
        Variant Calling Workflow
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../2_AutomaticVariantC/" class="md-nav__link">
        Automating a Variant Calling Workflow
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          RNA-seq Mapping And Count Data Workflow
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        RNA-seq Mapping And Count Data Workflow
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#assumptions" class="md-nav__link">
    Assumptions
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#alignment-to-a-reference-genome" class="md-nav__link">
    Alignment to a reference genome
  </a>
  
    <nav class="md-nav" aria-label="Alignment to a reference genome">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#preparation-of-the-genome" class="md-nav__link">
    Preparation of the genome
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#alignment-on-the-genome" class="md-nav__link">
    Alignment on the genome
  </a>
  
    <nav class="md-nav" aria-label="Alignment on the genome">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#converting-sam-files-to-bam-files" class="md-nav__link">
    Converting SAM files to BAM files
  </a>
  
    <nav class="md-nav" aria-label="Converting SAM files to BAM files">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#a-quick-look-into-the-sam-file" class="md-nav__link">
    A quick look into the sam file
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#some-stats-on-your-mapping" class="md-nav__link">
    Some stats on your mapping:
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#read-summarization" class="md-nav__link">
    Read Summarization
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#counting" class="md-nav__link">
    Counting
  </a>
  
    <nav class="md-nav" aria-label="Counting">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#subread-and-featurecounts" class="md-nav__link">
    Subread and FeatureCounts
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#group-exercise" class="md-nav__link">
    Group Exercise
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../4_IntroductiontoHPC/" class="md-nav__link">
        Introduction to HPC
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../5_working_with_job_scheduler/" class="md-nav__link">
        Working with job scheduler
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../6_supplementary_1/" class="md-nav__link">
        S1 : Accessing software via modules
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../7_supplementary_2/" class="md-nav__link">
        S2 : slurm profiling
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../8_supplementary_3/" class="md-nav__link">
        S3 : Solutions
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#assumptions" class="md-nav__link">
    Assumptions
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#alignment-to-a-reference-genome" class="md-nav__link">
    Alignment to a reference genome
  </a>
  
    <nav class="md-nav" aria-label="Alignment to a reference genome">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#preparation-of-the-genome" class="md-nav__link">
    Preparation of the genome
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#alignment-on-the-genome" class="md-nav__link">
    Alignment on the genome
  </a>
  
    <nav class="md-nav" aria-label="Alignment on the genome">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#converting-sam-files-to-bam-files" class="md-nav__link">
    Converting SAM files to BAM files
  </a>
  
    <nav class="md-nav" aria-label="Converting SAM files to BAM files">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#a-quick-look-into-the-sam-file" class="md-nav__link">
    A quick look into the sam file
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#some-stats-on-your-mapping" class="md-nav__link">
    Some stats on your mapping:
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#read-summarization" class="md-nav__link">
    Read Summarization
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#counting" class="md-nav__link">
    Counting
  </a>
  
    <nav class="md-nav" aria-label="Counting">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#subread-and-featurecounts" class="md-nav__link">
    Subread and FeatureCounts
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#group-exercise" class="md-nav__link">
    Group Exercise
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                
  <a href="https://github.com/GenomicsAotearoa/Workshop-Bash_Scripting_And_HPC_Job_Scheduler.git/edit/master/docs/3_RNAseq.md" title="Edit this page" class="md-content__button md-icon">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25Z"/></svg>
  </a>



<h1 id="rna-seq-mapping-and-count-data-workflow">RNA-seq Mapping And Count Data Workflow<a class="headerlink" href="#rna-seq-mapping-and-count-data-workflow" title="Permanent link">&para;</a></h1>
<p style="text-align:left;">
    <b><a class="btn" href="https://genomicsaotearoa.github.io/Workshop-Bash_Scripting_And_HPC_Job_Scheduler/workshop_material/2_AutomaticVariantC.html" style="background: var(--bs-green);font-weight:bold">&laquo;2. Automating a Variant Calling Workflow</a></b>
    <span style="float:right;">
     <b><a class="btn" href="https://genomicsaotearoa.github.io/Workshop-Bash_Scripting_And_HPC_Job_Scheduler/workshop_material/4_IntroductiontoHPC.html" style="background: var(--bs-green);font-weight:bold">4. Introduction to HPC  &raquo;</a></b>
    </span>
</p>

<p>This material is extracted from the <a href="https://github.com/GenomicsAotearoa/RNA-seq-workshop">RNA-seq workshop</a> Lesson</p>
<div class="admonition abstract">
<p class="admonition-title">Aim</p>
<ul>
<li>To develop a pipeline that does mapping and count the number of reads that mapped then overall put all these steps into a script.</li>
</ul>
</div>
<div class="admonition abstract">
<p class="admonition-title">Objectives and overall workflow</p>
<ul>
<li>Understand and perform the steps involved in RNA-seq mapping and read count.</li>
<li>Use command line tools to run the pipeline.</li>
</ul>
</div>
<p><center>
<img alt="image" src="../nesi_images/rnaseq_workflowdiagram.png" />
</center></p>
<h3 id="assumptions">Assumptions<a class="headerlink" href="#assumptions" title="Permanent link">&para;</a></h3>
<ul>
<li>You have already performed trimming and filtering of your reads and saved in a directory called trimmed_reads.</li>
<li>You have a reference genome saved in a directory called ref_genome.</li>
</ul>
<p>In this workshop, we have already trimmed the reads and downloaded the reference genome for you.
First, it is always good to verify where we are:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>$ <span class="nb">cd</span> ~/scripting_workshop
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>$ <span class="nb">pwd</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>/home/<span class="o">[</span>your_username<span class="o">]</span>/scripting_workshop
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="c1"># good I am ready to work</span>
</code></pre></div>
<p>Checking to make sure we have the directory and files for the workshop.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>$ ls
<a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a>rna_seq  variant_calling scheduler
</code></pre></div>
<blockquote>
<p>hint : If you do not have the workshop directory, you can copy it using the command: <code>cp -r  /nesi/project/nesi02659/scripting_workshop/ ~</code>  </p>
</blockquote>
<div class="highlight"><pre><span></span><code><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>$ <span class="nb">cd</span> rna_seq
<a id="__codelineno-2-2" name="__codelineno-2-2" href="#__codelineno-2-2"></a>
<a id="__codelineno-2-3" name="__codelineno-2-3" href="#__codelineno-2-3"></a>$ ls
<a id="__codelineno-2-4" name="__codelineno-2-4" href="#__codelineno-2-4"></a>ref_genome  trimmed_reads 
</code></pre></div>
<h2 id="alignment-to-a-reference-genome">Alignment to a reference genome<a class="headerlink" href="#alignment-to-a-reference-genome" title="Permanent link">&para;</a></h2>
<p>RNA-seq generate gene expression information by quantifying the number of transcripts (per gene) in a sample. This is acompished by counting the number of transcripts that have been sequenced - the more active a gene is, the more transcripts will be in a sample, and the more reads will be generated from that transcript.</p>
<p>For RNA-seq, we need to align or map each read back to the genome, to see which gene produced it.
- Highly expressed genes will generate lots of transcripts, so there will be lots of reads that map back to the position of that transcript in the genome.
- The per-gene data we work with in an RNA-seq experiment are counts: the number of reads from each sample that originated from that gene.</p>
<h3 id="preparation-of-the-genome">Preparation of the genome<a class="headerlink" href="#preparation-of-the-genome" title="Permanent link">&para;</a></h3>
<p>To be able to map (align) sequencing reads on the genome, the genome needs to be indexed first. In this workshop we will use <a href="https://www.nature.com/articles/nmeth.3317">HISAT2</a>.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a>$ <span class="nb">cd</span> ~/scripting_workshop/rna_seq/ref_genome
<a id="__codelineno-3-2" name="__codelineno-3-2" href="#__codelineno-3-2"></a>
<a id="__codelineno-3-3" name="__codelineno-3-3" href="#__codelineno-3-3"></a><span class="c1">#to list what is in your directory:</span>
<a id="__codelineno-3-4" name="__codelineno-3-4" href="#__codelineno-3-4"></a>$ ls ~/scripting_workshop/rna_seq/ref_genome
<a id="__codelineno-3-5" name="__codelineno-3-5" href="#__codelineno-3-5"></a>Saccharomyces_cerevisiae.R64-1-1.99.gtf  Saccharomyces_cerevisiae.R64-1-1.dna.toplevel.fa
<a id="__codelineno-3-6" name="__codelineno-3-6" href="#__codelineno-3-6"></a>
<a id="__codelineno-3-7" name="__codelineno-3-7" href="#__codelineno-3-7"></a>$ module load HISAT2/2.2.0-gimkl-2020a
<a id="__codelineno-3-8" name="__codelineno-3-8" href="#__codelineno-3-8"></a>
<a id="__codelineno-3-9" name="__codelineno-3-9" href="#__codelineno-3-9"></a><span class="c1"># index file:</span>
<a id="__codelineno-3-10" name="__codelineno-3-10" href="#__codelineno-3-10"></a>$ hisat2-build -p <span class="m">4</span> -f Saccharomyces_cerevisiae.R64-1-1.dna.toplevel.fa Saccharomyces_cerevisiae.R64-1-1.dna.toplevel
<a id="__codelineno-3-11" name="__codelineno-3-11" href="#__codelineno-3-11"></a>
<a id="__codelineno-3-12" name="__codelineno-3-12" href="#__codelineno-3-12"></a><span class="c1">#list what is in the directory:</span>
<a id="__codelineno-3-13" name="__codelineno-3-13" href="#__codelineno-3-13"></a>$ ls ~/scripting_workshop/rna_seq/ref_genome
<a id="__codelineno-3-14" name="__codelineno-3-14" href="#__codelineno-3-14"></a>Saccharomyces_cerevisiae.R64-1-1.99.gtf              Saccharomyces_cerevisiae.R64-1-1.dna.toplevel.4.ht2  Saccharomyces_cerevisiae.R64-1-1.dna.toplevel.8.ht2
<a id="__codelineno-3-15" name="__codelineno-3-15" href="#__codelineno-3-15"></a>Saccharomyces_cerevisiae.R64-1-1.dna.toplevel.1.ht2  Saccharomyces_cerevisiae.R64-1-1.dna.toplevel.5.ht2  Saccharomyces_cerevisiae.R64-1-1.dna.toplevel.fa
<a id="__codelineno-3-16" name="__codelineno-3-16" href="#__codelineno-3-16"></a>Saccharomyces_cerevisiae.R64-1-1.dna.toplevel.2.ht2  Saccharomyces_cerevisiae.R64-1-1.dna.toplevel.6.ht2
<a id="__codelineno-3-17" name="__codelineno-3-17" href="#__codelineno-3-17"></a>Saccharomyces_cerevisiae.R64-1-1.dna.toplevel.3.ht2  Saccharomyces_cerevisiae.R64-1-1.dna.toplevel.7.ht2
</code></pre></div>
<blockquote>
<p><strong>Arguments:</strong>
 * <strong>-p</strong> number of threads
 * <strong>-f</strong> fasta file</p>
</blockquote>
<p>How many files were created during the indexing process?</p>
<h2 id="alignment-on-the-genome">Alignment on the genome<a class="headerlink" href="#alignment-on-the-genome" title="Permanent link">&para;</a></h2>
<p>Now that the genome is prepared. Sequencing reads can be aligned.</p>
<p>Information required:</p>
<ul>
<li>Where the sequence information is stored (e.g. fastq files ...) ?</li>
<li>What kind of sequencing: Single End or Paired end ?</li>
<li>Where are stored the indexes and the genome? </li>
<li>
<p>Where will the mapping files be stored?</p>
</li>
<li>
<p>Now, lets move one folder up (into the rna_seq folder):</p>
</li>
</ul>
<div class="highlight"><pre><span></span><code><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a>$ <span class="nb">cd</span> ..
<a id="__codelineno-4-2" name="__codelineno-4-2" href="#__codelineno-4-2"></a>
<a id="__codelineno-4-3" name="__codelineno-4-3" href="#__codelineno-4-3"></a>$ ls
<a id="__codelineno-4-4" name="__codelineno-4-4" href="#__codelineno-4-4"></a>ref_genome  trimmed_reads
</code></pre></div>
<p>Let's map one of our sample to the reference genome</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a>$ <span class="nb">pwd</span>
<a id="__codelineno-5-2" name="__codelineno-5-2" href="#__codelineno-5-2"></a>/home/<span class="o">[</span>Your_Username<span class="o">]</span>/scripting_workshop/rna_seq/
<a id="__codelineno-5-3" name="__codelineno-5-3" href="#__codelineno-5-3"></a>
<a id="__codelineno-5-4" name="__codelineno-5-4" href="#__codelineno-5-4"></a>
<a id="__codelineno-5-5" name="__codelineno-5-5" href="#__codelineno-5-5"></a>$ mkdir Mapping
<a id="__codelineno-5-6" name="__codelineno-5-6" href="#__codelineno-5-6"></a>
<a id="__codelineno-5-7" name="__codelineno-5-7" href="#__codelineno-5-7"></a>$ ls
<a id="__codelineno-5-8" name="__codelineno-5-8" href="#__codelineno-5-8"></a>ref_genome  Mapping  trimmed_reads
</code></pre></div>
<p>let's use a for loop to process our samples:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a>$ <span class="nb">cd</span> trimmed_reads
<a id="__codelineno-6-2" name="__codelineno-6-2" href="#__codelineno-6-2"></a>
<a id="__codelineno-6-3" name="__codelineno-6-3" href="#__codelineno-6-3"></a>$ ls
<a id="__codelineno-6-4" name="__codelineno-6-4" href="#__codelineno-6-4"></a>SRR014335-chr1.fastq  SRR014336-chr1.fastq  SRR014337-chr1.fastq  SRR014339-chr1.fastq  SRR014340-chr1.fastq  SRR014341-chr1.fastq
<a id="__codelineno-6-5" name="__codelineno-6-5" href="#__codelineno-6-5"></a>
<a id="__codelineno-6-6" name="__codelineno-6-6" href="#__codelineno-6-6"></a>$ <span class="k">for</span> filename <span class="k">in</span> *
<a id="__codelineno-6-7" name="__codelineno-6-7" href="#__codelineno-6-7"></a>&gt; <span class="k">do</span>
<a id="__codelineno-6-8" name="__codelineno-6-8" href="#__codelineno-6-8"></a>&gt; <span class="nv">base</span><span class="o">=</span><span class="k">$(</span>basename <span class="si">${</span><span class="nv">filename</span><span class="si">}</span> .fastq<span class="k">)</span>
<a id="__codelineno-6-9" name="__codelineno-6-9" href="#__codelineno-6-9"></a>&gt; hisat2 -p <span class="m">4</span> -x ../ref_genome/Saccharomyces_cerevisiae.R64-1-1.dna.toplevel -U <span class="nv">$filename</span> -S ../Mapping/<span class="si">${</span><span class="nv">base</span><span class="si">}</span>.sam --summary-file ../Mapping/<span class="si">${</span><span class="nv">base</span><span class="si">}</span>_summary.txt
<a id="__codelineno-6-10" name="__codelineno-6-10" href="#__codelineno-6-10"></a>&gt; <span class="k">done</span>
</code></pre></div>
<blockquote>
<p><strong>Arguments:</strong>
 * <strong>-x</strong> The basename of the index for the reference genome. 
 * <strong>-U</strong> Comma-separated list of files containing unpaired reads to be aligned
 * <strong>-S</strong> File to write SAM alignments to. By default, alignments are written to the “standard out” or “stdout” filehandle  </p>
</blockquote>
<p>Now we can explore our SAM files.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a>$ <span class="nb">cd</span> ../Mapping
<a id="__codelineno-7-2" name="__codelineno-7-2" href="#__codelineno-7-2"></a>
<a id="__codelineno-7-3" name="__codelineno-7-3" href="#__codelineno-7-3"></a>$ ls
<a id="__codelineno-7-4" name="__codelineno-7-4" href="#__codelineno-7-4"></a>SRR014335-chr1.sam          SRR014336-chr1_summary.txt  SRR014339-chr1.sam          SRR014340-chr1_summary.txt
<a id="__codelineno-7-5" name="__codelineno-7-5" href="#__codelineno-7-5"></a>SRR014335-chr1_summary.txt  SRR014337-chr1.sam          SRR014339-chr1_summary.txt  SRR014341-chr1.sam
<a id="__codelineno-7-6" name="__codelineno-7-6" href="#__codelineno-7-6"></a>SRR014336-chr1.sam          SRR014337-chr1_summary.txt  SRR014340-chr1.sam          SRR014341-chr1_summary.txt
</code></pre></div>
<h3 id="converting-sam-files-to-bam-files">Converting SAM files to BAM files<a class="headerlink" href="#converting-sam-files-to-bam-files" title="Permanent link">&para;</a></h3>
<p>The SAM file, is a tab-delimited text file that contains information for each individual read and its alignment to the genome. While we do not have time to go into detail about the features of the SAM format, the paper by Heng Li et al. provides a lot more detail on the specification.</p>
<p>The compressed binary version of SAM is called a BAM file. We use this version to reduce size and to allow for indexing, which enables efficient random access of the data contained within the file.</p>
<h4 id="a-quick-look-into-the-sam-file">A quick look into the sam file<a class="headerlink" href="#a-quick-look-into-the-sam-file" title="Permanent link">&para;</a></h4>
<div class="highlight"><pre><span></span><code><a id="__codelineno-8-1" name="__codelineno-8-1" href="#__codelineno-8-1"></a>$ less SRR014335-chr1.sam 
<a id="__codelineno-8-2" name="__codelineno-8-2" href="#__codelineno-8-2"></a>
<a id="__codelineno-8-3" name="__codelineno-8-3" href="#__codelineno-8-3"></a>The file begins with a header, which is optional. The header is used to describe the <span class="nb">source</span> of data, reference sequence, method of alignment, etc., this will change depending on the aligner being used. Following the header is the alignment section. Each line that follows corresponds to alignment information <span class="k">for</span> a single read. Each alignment line has <span class="m">11</span> mandatory fields <span class="k">for</span> essential mapping information and a variable number of other fields <span class="k">for</span> aligner specific information. An example entry from a SAM file is displayed below with the different fields highlighted.
</code></pre></div>
<p>We will convert the SAM file to BAM format using the samtools program with the view command and tell this command that the input is in SAM format (<code>-S</code>) and to output BAM format (<code>-b</code>):</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-9-1" name="__codelineno-9-1" href="#__codelineno-9-1"></a>$ module load SAMtools/1.10-GCC-9.2.0
<a id="__codelineno-9-2" name="__codelineno-9-2" href="#__codelineno-9-2"></a>
<a id="__codelineno-9-3" name="__codelineno-9-3" href="#__codelineno-9-3"></a>$ <span class="k">for</span> filename <span class="k">in</span> *.sam
<a id="__codelineno-9-4" name="__codelineno-9-4" href="#__codelineno-9-4"></a>&gt; <span class="k">do</span>
<a id="__codelineno-9-5" name="__codelineno-9-5" href="#__codelineno-9-5"></a>&gt; <span class="nv">base</span><span class="o">=</span><span class="k">$(</span>basename <span class="si">${</span><span class="nv">filename</span><span class="si">}</span> .sam<span class="k">)</span>
<a id="__codelineno-9-6" name="__codelineno-9-6" href="#__codelineno-9-6"></a>&gt; samtools view -S -b <span class="si">${</span><span class="nv">filename</span><span class="si">}</span> -o <span class="si">${</span><span class="nv">base</span><span class="si">}</span>.bam
<a id="__codelineno-9-7" name="__codelineno-9-7" href="#__codelineno-9-7"></a>&gt; <span class="k">done</span>
<a id="__codelineno-9-8" name="__codelineno-9-8" href="#__codelineno-9-8"></a>
<a id="__codelineno-9-9" name="__codelineno-9-9" href="#__codelineno-9-9"></a>$ ls
<a id="__codelineno-9-10" name="__codelineno-9-10" href="#__codelineno-9-10"></a>SRR014335-chr1.bam  SRR014336-chr1.bam  SRR014337-chr1.bam  SRR014339-chr1.bam  SRR014340-chr1.bam  SRR014341-chr1.bam
<a id="__codelineno-9-11" name="__codelineno-9-11" href="#__codelineno-9-11"></a>SRR014335-chr1.sam  SRR014336-chr1.sam  SRR014337-chr1.sam  SRR014339-chr1.sam  SRR014340-chr1.sam  SRR014341-chr1.sam
</code></pre></div>
<p>Next we sort the BAM file using the sort command from samtools. <code>-o</code> tells the command where to write the output.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>SAM/BAM files can be sorted in multiple ways, e.g. by location of alignment on the chromosome, by read name, etc. It is important to be aware that different alignment tools will output differently sorted SAM/BAM, and different downstream tools require differently sorted alignment files as input.**</p>
</div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-10-1" name="__codelineno-10-1" href="#__codelineno-10-1"></a>$ <span class="k">for</span> filename <span class="k">in</span> *.bam
<a id="__codelineno-10-2" name="__codelineno-10-2" href="#__codelineno-10-2"></a>&gt; <span class="k">do</span>
<a id="__codelineno-10-3" name="__codelineno-10-3" href="#__codelineno-10-3"></a>&gt; <span class="nv">base</span><span class="o">=</span><span class="k">$(</span>basename <span class="si">${</span><span class="nv">filename</span><span class="si">}</span> .bam<span class="k">)</span>
<a id="__codelineno-10-4" name="__codelineno-10-4" href="#__codelineno-10-4"></a>&gt; samtools sort -o <span class="si">${</span><span class="nv">base</span><span class="si">}</span>_sorted.bam <span class="si">${</span><span class="nv">filename</span><span class="si">}</span>
<a id="__codelineno-10-5" name="__codelineno-10-5" href="#__codelineno-10-5"></a>&gt; <span class="k">done</span>
</code></pre></div>
<p>We can use samtools to learn more about the bam file as well.</p>
<h4 id="some-stats-on-your-mapping">Some stats on your mapping:<a class="headerlink" href="#some-stats-on-your-mapping" title="Permanent link">&para;</a></h4>
<div class="highlight"><pre><span></span><code><a id="__codelineno-11-1" name="__codelineno-11-1" href="#__codelineno-11-1"></a>$ samtools flagstat SRR014335-chr1_sorted.bam 
<a id="__codelineno-11-2" name="__codelineno-11-2" href="#__codelineno-11-2"></a><span class="m">156984</span> + <span class="m">0</span> <span class="k">in</span> total <span class="o">(</span>QC-passed reads + QC-failed reads<span class="o">)</span>
<a id="__codelineno-11-3" name="__codelineno-11-3" href="#__codelineno-11-3"></a><span class="m">31894</span> + <span class="m">0</span> secondary
<a id="__codelineno-11-4" name="__codelineno-11-4" href="#__codelineno-11-4"></a><span class="m">0</span> + <span class="m">0</span> supplementary
<a id="__codelineno-11-5" name="__codelineno-11-5" href="#__codelineno-11-5"></a><span class="m">0</span> + <span class="m">0</span> duplicates
<a id="__codelineno-11-6" name="__codelineno-11-6" href="#__codelineno-11-6"></a><span class="m">136447</span> + <span class="m">0</span> mapped <span class="o">(</span><span class="m">86</span>.92% : N/A<span class="o">)</span>
<a id="__codelineno-11-7" name="__codelineno-11-7" href="#__codelineno-11-7"></a><span class="m">0</span> + <span class="m">0</span> paired <span class="k">in</span> sequencing
<a id="__codelineno-11-8" name="__codelineno-11-8" href="#__codelineno-11-8"></a><span class="m">0</span> + <span class="m">0</span> read1
<a id="__codelineno-11-9" name="__codelineno-11-9" href="#__codelineno-11-9"></a><span class="m">0</span> + <span class="m">0</span> read2
<a id="__codelineno-11-10" name="__codelineno-11-10" href="#__codelineno-11-10"></a><span class="m">0</span> + <span class="m">0</span> properly paired <span class="o">(</span>N/A : N/A<span class="o">)</span>
<a id="__codelineno-11-11" name="__codelineno-11-11" href="#__codelineno-11-11"></a><span class="m">0</span> + <span class="m">0</span> with itself and mate mapped
<a id="__codelineno-11-12" name="__codelineno-11-12" href="#__codelineno-11-12"></a><span class="m">0</span> + <span class="m">0</span> singletons <span class="o">(</span>N/A : N/A<span class="o">)</span>
<a id="__codelineno-11-13" name="__codelineno-11-13" href="#__codelineno-11-13"></a><span class="m">0</span> + <span class="m">0</span> with mate mapped to a different chr
<a id="__codelineno-11-14" name="__codelineno-11-14" href="#__codelineno-11-14"></a><span class="m">0</span> + <span class="m">0</span> with mate mapped to a different chr <span class="o">(</span>mapQ&gt;<span class="o">=</span><span class="m">5</span><span class="o">)</span>
</code></pre></div>
<hr />
<h2 id="read-summarization">Read Summarization<a class="headerlink" href="#read-summarization" title="Permanent link">&para;</a></h2>
<p>Sequencing reads often need to be assigned to genomic features of interest after they are mapped to the reference genome. This process is often called read summarization or read quantification. Read summarization is required by a number of downstream analyses such as gene expression analysis and histone modification analysis. The output of read summarization
is a count table, in which the number of reads assigned to each feature in each library is recorded.</p>
<h2 id="counting">Counting<a class="headerlink" href="#counting" title="Permanent link">&para;</a></h2>
<ul>
<li>We need to do some counting!</li>
<li>Want to generate count data for each gene (actually each exon) - how many reads mapped to each exon in the genome, from each of our samples?</li>
<li>Once we have that information, we can start thinking about how to determine which genes were differentially expressed in our study.</li>
</ul>
<h3 id="subread-and-featurecounts">Subread and FeatureCounts<a class="headerlink" href="#subread-and-featurecounts" title="Permanent link">&para;</a></h3>
<ul>
<li>The featureCounts tool from the Subread package can be used to count how many reads aligned to each genome feature (exon).</li>
<li>Need to specify the annotation informatyion (.gtf file) 
You can process all the samples at once:</li>
</ul>
<div class="highlight"><pre><span></span><code><a id="__codelineno-12-1" name="__codelineno-12-1" href="#__codelineno-12-1"></a>$ <span class="nb">cd</span> ~/scripting_workshop/rna_seq
<a id="__codelineno-12-2" name="__codelineno-12-2" href="#__codelineno-12-2"></a>
<a id="__codelineno-12-3" name="__codelineno-12-3" href="#__codelineno-12-3"></a>$ module load Subread/2.0.0-GCC-9.2.0
<a id="__codelineno-12-4" name="__codelineno-12-4" href="#__codelineno-12-4"></a>
<a id="__codelineno-12-5" name="__codelineno-12-5" href="#__codelineno-12-5"></a>$ <span class="nb">pwd</span>
<a id="__codelineno-12-6" name="__codelineno-12-6" href="#__codelineno-12-6"></a>/home/<span class="o">[</span>Your_Username<span class="o">]</span>/scripting_workshop/rna_seq
<a id="__codelineno-12-7" name="__codelineno-12-7" href="#__codelineno-12-7"></a>
<a id="__codelineno-12-8" name="__codelineno-12-8" href="#__codelineno-12-8"></a>$ mkdir Counts
<a id="__codelineno-12-9" name="__codelineno-12-9" href="#__codelineno-12-9"></a>
<a id="__codelineno-12-10" name="__codelineno-12-10" href="#__codelineno-12-10"></a>$ <span class="nb">cd</span> Counts
<a id="__codelineno-12-11" name="__codelineno-12-11" href="#__codelineno-12-11"></a>
<a id="__codelineno-12-12" name="__codelineno-12-12" href="#__codelineno-12-12"></a>$ featureCounts -a ../ref_genome/Saccharomyces_cerevisiae.R64-1-1.99.gtf -o ./yeast_counts.txt -T <span class="m">2</span> -t exon -g gene_id ../Mapping/*sorted.bam
</code></pre></div>
<div class="admonition abstract">
<p class="admonition-title"><strong>Arguments:</strong></p>
<ul>
<li><strong>-a</strong> Name of an annotation file. GTF/GFF format by default</li>
<li><strong>-o</strong> Name of output file including read counts</li>
<li><strong>-T</strong> Specify the number of threads/CPUs used for mapping. 1 by default</li>
<li><strong>-t</strong> Specify feature type in GTF annotation. 'exon' by default. Features used for read counting will be extracted from annotation using the provided value.</li>
<li><strong>-g</strong> Specify attribute type in GTF annotation. 'gene_id' by default. Meta-features used for read counting will be extracted from annotation using the provided value.</li>
</ul>
</div>
<hr />
<h2 id="group-exercise">Group Exercise<a class="headerlink" href="#group-exercise" title="Permanent link">&para;</a></h2>
<p>Now, let's work together in our groups to create an RNA-seq mapping and count script.</p>
<hr />
<p>at this stage we have mastered the art of writing scripts, instead of running them on the commandline, let us now run them on HPC.</p>

              
            </article>
            
          </div>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    
    <nav class="md-footer__inner md-grid" aria-label="Footer" >
      
        
        <a href="../2_AutomaticVariantC/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Automating a Variant Calling Workflow" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Automating a Variant Calling Workflow
            </div>
          </div>
        </a>
      
      
        
        <a href="../4_IntroductiontoHPC/" class="md-footer__link md-footer__link--next" aria-label="Next: Introduction to HPC" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Introduction to HPC
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "..", "features": ["navigation.tracking", "navigation.sections", "navigation.expand"], "search": "../assets/javascripts/workers/search.b97dbffb.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.6c7ad80a.min.js"></script>
      
    
  </body>
</html>