
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
        <meta name="author" content="Genomics Aotearoa & NeSI">
      
      
        <link rel="canonical" href="https://genomicsaotearoa.github.io/Workshop-Bash_Scripting_And_HPC_Job_Scheduler/8_supplementary_3/">
      
      
        <link rel="prev" href="../7_supplementary_2/">
      
      
      
      <link rel="icon" href="../nesi_images/nesi_ga.png">
      <meta name="generator" content="mkdocs-1.6.0, mkdocs-material-9.5.32">
    
    
      
        <title>S3 : Solutions - Introduction to Bash Scripting and HPC Scheduler</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.3cba04c6.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Mukta:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Mukta";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../stylesheets/extra.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
   <link href="../assets/stylesheets/glightbox.min.css" rel="stylesheet"/><style>
    html.glightbox-open { overflow: initial; height: 100%; }
    .gslide-title { margin-top: 0px; user-select: text; }
    .gslide-desc { color: #666; user-select: text; }
    .gslide-image img { background: white; }
    .gscrollbar-fixer { padding-right: 15px; }
    .gdesc-inner { font-size: 0.75rem; }
    body[data-md-color-scheme="slate"] .gdesc-inner { background: var(--md-default-bg-color);}
    body[data-md-color-scheme="slate"] .gslide-title { color: var(--md-default-fg-color);}
    body[data-md-color-scheme="slate"] .gslide-desc { color: var(--md-default-fg-color);}</style> <script src="../assets/javascripts/glightbox.min.js"></script></head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#s3-solutions" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Introduction to Bash Scripting and HPC Scheduler" class="md-header__button md-logo" aria-label="Introduction to Bash Scripting and HPC Scheduler" data-md-component="logo">
      
  <img src="../nesi_images/nesi_ga.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Introduction to Bash Scripting and HPC Scheduler
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              S3 : Solutions
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme)" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m14.3 16-.7-2h-3.2l-.7 2H7.8L11 7h2l3.2 9h-1.9M20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69m-9.15 3.96h2.3L12 9l-1.15 3.65Z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="custom" data-md-color-accent="indigo"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_2">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/GenomicsAotearoa/Workshop-Bash_Scripting_And_HPC_Job_Scheduler" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GenomicsAotearoa/Workshop-Bash_Scripting_And_HPC_Job_Scheduler
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Introduction to Bash Scripting and HPC Scheduler" class="md-nav__button md-logo" aria-label="Introduction to Bash Scripting and HPC Scheduler" data-md-component="logo">
      
  <img src="../nesi_images/nesi_ga.png" alt="logo">

    </a>
    Introduction to Bash Scripting and HPC Scheduler
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/GenomicsAotearoa/Workshop-Bash_Scripting_And_HPC_Job_Scheduler" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GenomicsAotearoa/Workshop-Bash_Scripting_And_HPC_Job_Scheduler
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../0_Background/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Background
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../1_DesigningVariantC/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Variant Calling Workflow
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../2_AutomaticVariantC/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Automating a Variant Calling Workflow
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../3_RNAseq/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    RNA-Seq Mapping And Count Data Workflow
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../4_IntroductiontoHPC/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction to HPC
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../5_working_with_job_scheduler/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Working with job scheduler
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../6_supplementary_1/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    S1 : Accessing software via modules
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../7_supplementary_2/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    S2 : slurm profiling
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    S3 : Solutions
  </span>
  

      </a>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


<h1 id="s3-solutions">S3 : Solutions<a class="headerlink" href="#s3-solutions" title="Permanent link">&para;</a></h1>
<details class="circle-check">
<summary>RNA-Seq Mapping and Count Data 😺</summary>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-1"> 1</a></span>
<span class="normal"><a href="#__codelineno-0-2"> 2</a></span>
<span class="normal"><a href="#__codelineno-0-3"> 3</a></span>
<span class="normal"><a href="#__codelineno-0-4"> 4</a></span>
<span class="normal"><a href="#__codelineno-0-5"> 5</a></span>
<span class="normal"><a href="#__codelineno-0-6"> 6</a></span>
<span class="normal"><a href="#__codelineno-0-7"> 7</a></span>
<span class="normal"><a href="#__codelineno-0-8"> 8</a></span>
<span class="normal"><a href="#__codelineno-0-9"> 9</a></span>
<span class="normal"><a href="#__codelineno-0-10">10</a></span>
<span class="normal"><a href="#__codelineno-0-11">11</a></span>
<span class="normal"><a href="#__codelineno-0-12">12</a></span>
<span class="normal"><a href="#__codelineno-0-13">13</a></span>
<span class="normal"><a href="#__codelineno-0-14">14</a></span>
<span class="normal"><a href="#__codelineno-0-15">15</a></span>
<span class="normal"><a href="#__codelineno-0-16">16</a></span>
<span class="normal"><a href="#__codelineno-0-17">17</a></span>
<span class="normal"><a href="#__codelineno-0-18">18</a></span>
<span class="normal"><a href="#__codelineno-0-19">19</a></span>
<span class="normal"><a href="#__codelineno-0-20">20</a></span>
<span class="normal"><a href="#__codelineno-0-21">21</a></span>
<span class="normal"><a href="#__codelineno-0-22">22</a></span>
<span class="normal"><a href="#__codelineno-0-23">23</a></span>
<span class="normal"><a href="#__codelineno-0-24">24</a></span>
<span class="normal"><a href="#__codelineno-0-25">25</a></span>
<span class="normal"><a href="#__codelineno-0-26">26</a></span>
<span class="normal"><a href="#__codelineno-0-27">27</a></span>
<span class="normal"><a href="#__codelineno-0-28">28</a></span>
<span class="normal"><a href="#__codelineno-0-29">29</a></span>
<span class="normal"><a href="#__codelineno-0-30">30</a></span>
<span class="normal"><a href="#__codelineno-0-31">31</a></span>
<span class="normal"><a href="#__codelineno-0-32">32</a></span>
<span class="normal"><a href="#__codelineno-0-33">33</a></span>
<span class="normal"><a href="#__codelineno-0-34">34</a></span>
<span class="normal"><a href="#__codelineno-0-35">35</a></span>
<span class="normal"><a href="#__codelineno-0-36">36</a></span>
<span class="normal"><a href="#__codelineno-0-37">37</a></span>
<span class="normal"><a href="#__codelineno-0-38">38</a></span>
<span class="normal"><a href="#__codelineno-0-39">39</a></span>
<span class="normal"><a href="#__codelineno-0-40">40</a></span>
<span class="normal"><a href="#__codelineno-0-41">41</a></span>
<span class="normal"><a href="#__codelineno-0-42">42</a></span>
<span class="normal"><a href="#__codelineno-0-43">43</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1"></a><span class="ch">#!/bin/bash -e</span>
<a id="__codelineno-0-2" name="__codelineno-0-2"></a>
<a id="__codelineno-0-3" name="__codelineno-0-3"></a><span class="c1"># Jane Doe</span>
<a id="__codelineno-0-4" name="__codelineno-0-4"></a><span class="c1"># 13 April 2023</span>
<a id="__codelineno-0-5" name="__codelineno-0-5"></a><span class="c1">#This script uses relative paths. It has to be executed from rna_seq parent directory OR edit paths accordingly. </span>
<a id="__codelineno-0-6" name="__codelineno-0-6"></a>
<a id="__codelineno-0-7" name="__codelineno-0-7"></a><span class="c1"># Load required modules and setup the environment</span>
<a id="__codelineno-0-8" name="__codelineno-0-8"></a>module<span class="w"> </span>purge
<a id="__codelineno-0-9" name="__codelineno-0-9"></a>module<span class="w"> </span>load<span class="w"> </span>HISAT2/2.2.0-gimkl-2020a
<a id="__codelineno-0-10" name="__codelineno-0-10"></a>module<span class="w"> </span>load<span class="w"> </span>SAMtools/1.10-GCC-9.2.0
<a id="__codelineno-0-11" name="__codelineno-0-11"></a>module<span class="w"> </span>load<span class="w"> </span>Subread/2.0.0-GCC-9.2.0
<a id="__codelineno-0-12" name="__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13"></a><span class="c1">#Print the current working directory</span>
<a id="__codelineno-0-14" name="__codelineno-0-14"></a><span class="nb">echo</span><span class="w"> </span><span class="s2">&quot;</span><span class="si">${</span><span class="nv">PWD</span><span class="si">}</span><span class="s2">&quot;</span>
<a id="__codelineno-0-15" name="__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16"></a><span class="c1">#Create results directories. We are moving away from &quot;Mapping&quot; and &quot;Counts&quot;</span>
<a id="__codelineno-0-17" name="__codelineno-0-17"></a>mkdir<span class="w"> </span>-p<span class="w"> </span>results/<span class="o">{</span>sam,bam,counts<span class="o">}</span>
<a id="__codelineno-0-18" name="__codelineno-0-18"></a>
<a id="__codelineno-0-19" name="__codelineno-0-19"></a><span class="c1"># Index genome</span>
<a id="__codelineno-0-20" name="__codelineno-0-20"></a>hisat2-build<span class="w"> </span>-p<span class="w"> </span><span class="m">2</span><span class="w"> </span>-f<span class="w"> </span>ref_genome/Saccharomyces_cerevisiae.R64-1-1.dna.toplevel.fa<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-0-21" name="__codelineno-0-21"></a><span class="w">                     </span>ref_genome/Saccharomyces_cerevisiae.R64-1-1.dna.toplevel
<a id="__codelineno-0-22" name="__codelineno-0-22"></a>
<a id="__codelineno-0-23" name="__codelineno-0-23"></a><span class="c1"># Align to indexed genome</span>
<a id="__codelineno-0-24" name="__codelineno-0-24"></a><span class="k">for</span><span class="w"> </span>filename<span class="w"> </span><span class="k">in</span><span class="w"> </span>trimmed_reads/*.fastq
<a id="__codelineno-0-25" name="__codelineno-0-25"></a><span class="k">do</span>
<a id="__codelineno-0-26" name="__codelineno-0-26"></a><span class="w">      </span><span class="c1"># Extract base name</span>
<a id="__codelineno-0-27" name="__codelineno-0-27"></a><span class="w">      </span><span class="nv">base</span><span class="o">=</span><span class="k">$(</span>basename<span class="w"> </span><span class="si">${</span><span class="nv">filename</span><span class="si">}</span><span class="w"> </span>.fastq<span class="k">)</span>
<a id="__codelineno-0-28" name="__codelineno-0-28"></a>
<a id="__codelineno-0-29" name="__codelineno-0-29"></a><span class="w">      </span><span class="c1"># Align to the reference genome</span>
<a id="__codelineno-0-30" name="__codelineno-0-30"></a><span class="w">      </span>hisat2<span class="w"> </span>-p<span class="w"> </span><span class="m">2</span><span class="w"> </span>-x<span class="w"> </span>ref_genome/Saccharomyces_cerevisiae.R64-1-1.dna.toplevel<span class="w"> </span>-U<span class="w"> </span><span class="nv">$filename</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-0-31" name="__codelineno-0-31"></a><span class="w">      </span>-S<span class="w"> </span>results/sam/<span class="si">${</span><span class="nv">base</span><span class="si">}</span>.sam<span class="w"> </span>--summary-file<span class="w"> </span>results/sam/<span class="si">${</span><span class="nv">base</span><span class="si">}</span>.summary.txt
<a id="__codelineno-0-32" name="__codelineno-0-32"></a>
<a id="__codelineno-0-33" name="__codelineno-0-33"></a><span class="w">      </span><span class="c1"># Convert SAM to BAM</span>
<a id="__codelineno-0-34" name="__codelineno-0-34"></a><span class="w">      </span>samtools<span class="w"> </span>view<span class="w"> </span>-S<span class="w"> </span>-b<span class="w"> </span>results/sam/<span class="si">${</span><span class="nv">base</span><span class="si">}</span>.sam<span class="w"> </span><span class="p">|</span><span class="w"> </span>samtools<span class="w"> </span>sort<span class="w"> </span>-o<span class="w"> </span>results/bam/<span class="si">${</span><span class="nv">base</span><span class="si">}</span>_sorted.bam
<a id="__codelineno-0-35" name="__codelineno-0-35"></a>
<a id="__codelineno-0-36" name="__codelineno-0-36"></a><span class="w">      </span><span class="c1"># Extract stats for mapping</span>
<a id="__codelineno-0-37" name="__codelineno-0-37"></a><span class="w">      </span>samtools<span class="w"> </span>flagstat<span class="w"> </span>results/bam/<span class="si">${</span><span class="nv">base</span><span class="si">}</span>_sorted.bam<span class="w"> </span>&gt;<span class="w"> </span>results/bam/<span class="si">${</span><span class="nv">base</span><span class="si">}</span>_mapstat.txt
<a id="__codelineno-0-38" name="__codelineno-0-38"></a><span class="k">done</span>
<a id="__codelineno-0-39" name="__codelineno-0-39"></a>
<a id="__codelineno-0-40" name="__codelineno-0-40"></a><span class="c1"># count how many reads aligned to each genome feature (exon).</span>
<a id="__codelineno-0-41" name="__codelineno-0-41"></a>featureCounts<span class="w"> </span>-a<span class="w"> </span>ref_genome/Saccharomyces_cerevisiae.R64-1-1.99.gtf<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-0-42" name="__codelineno-0-42"></a><span class="w">              </span>-o<span class="w"> </span>results/counts/yeast_counts.txt<span class="w"> </span>-T<span class="w"> </span><span class="m">2</span><span class="w"> </span>-t<span class="w"> </span>exon<span class="w"> </span>-g<span class="w"> </span>gene_id<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-0-43" name="__codelineno-0-43"></a><span class="w">                 </span>results/bam/<span class="si">${</span><span class="nv">base</span><span class="si">}</span>_sorted.bam
</code></pre></div></td></tr></table></div>
</details>
<details class="circle-check">
<summary>Exercise 5.4 😊</summary>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-1-1"> 1</a></span>
<span class="normal"><a href="#__codelineno-1-2"> 2</a></span>
<span class="normal"><a href="#__codelineno-1-3"> 3</a></span>
<span class="normal"><a href="#__codelineno-1-4"> 4</a></span>
<span class="normal"><a href="#__codelineno-1-5"> 5</a></span>
<span class="normal"><a href="#__codelineno-1-6"> 6</a></span>
<span class="normal"><a href="#__codelineno-1-7"> 7</a></span>
<span class="normal"><a href="#__codelineno-1-8"> 8</a></span>
<span class="normal"><a href="#__codelineno-1-9"> 9</a></span>
<span class="normal"><a href="#__codelineno-1-10">10</a></span>
<span class="normal"><a href="#__codelineno-1-11">11</a></span>
<span class="normal"><a href="#__codelineno-1-12">12</a></span>
<span class="normal"><a href="#__codelineno-1-13">13</a></span>
<span class="normal"><a href="#__codelineno-1-14">14</a></span>
<span class="normal"><a href="#__codelineno-1-15">15</a></span>
<span class="normal"><a href="#__codelineno-1-16">16</a></span>
<span class="normal"><a href="#__codelineno-1-17">17</a></span>
<span class="normal"><a href="#__codelineno-1-18">18</a></span>
<span class="normal"><a href="#__codelineno-1-19">19</a></span>
<span class="normal"><a href="#__codelineno-1-20">20</a></span>
<span class="normal"><a href="#__codelineno-1-21">21</a></span>
<span class="normal"><a href="#__codelineno-1-22">22</a></span>
<span class="normal"><a href="#__codelineno-1-23">23</a></span>
<span class="normal"><a href="#__codelineno-1-24">24</a></span>
<span class="normal"><a href="#__codelineno-1-25">25</a></span>
<span class="normal"><a href="#__codelineno-1-26">26</a></span>
<span class="normal"><a href="#__codelineno-1-27">27</a></span>
<span class="normal"><a href="#__codelineno-1-28">28</a></span>
<span class="normal"><a href="#__codelineno-1-29">29</a></span>
<span class="normal"><a href="#__codelineno-1-30">30</a></span>
<span class="normal"><a href="#__codelineno-1-31">31</a></span>
<span class="normal"><a href="#__codelineno-1-32">32</a></span>
<span class="normal"><a href="#__codelineno-1-33">33</a></span>
<span class="normal"><a href="#__codelineno-1-34">34</a></span>
<span class="normal"><a href="#__codelineno-1-35">35</a></span>
<span class="normal"><a href="#__codelineno-1-36">36</a></span>
<span class="normal"><a href="#__codelineno-1-37">37</a></span>
<span class="normal"><a href="#__codelineno-1-38">38</a></span>
<span class="normal"><a href="#__codelineno-1-39">39</a></span>
<span class="normal"><a href="#__codelineno-1-40">40</a></span>
<span class="normal"><a href="#__codelineno-1-41">41</a></span>
<span class="normal"><a href="#__codelineno-1-42">42</a></span>
<span class="normal"><a href="#__codelineno-1-43">43</a></span>
<span class="normal"><a href="#__codelineno-1-44">44</a></span>
<span class="normal"><a href="#__codelineno-1-45">45</a></span>
<span class="normal"><a href="#__codelineno-1-46">46</a></span>
<span class="normal"><a href="#__codelineno-1-47">47</a></span>
<span class="normal"><a href="#__codelineno-1-48">48</a></span>
<span class="normal"><a href="#__codelineno-1-49">49</a></span>
<span class="normal"><a href="#__codelineno-1-50">50</a></span>
<span class="normal"><a href="#__codelineno-1-51">51</a></span>
<span class="normal"><a href="#__codelineno-1-52">52</a></span>
<span class="normal"><a href="#__codelineno-1-53">53</a></span>
<span class="normal"><a href="#__codelineno-1-54">54</a></span>
<span class="normal"><a href="#__codelineno-1-55">55</a></span>
<span class="normal"><a href="#__codelineno-1-56">56</a></span>
<span class="normal"><a href="#__codelineno-1-57">57</a></span>
<span class="normal"><a href="#__codelineno-1-58">58</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1"></a><span class="ch">#!/bin/bash -e</span>
<a id="__codelineno-1-2" name="__codelineno-1-2"></a>
<a id="__codelineno-1-3" name="__codelineno-1-3"></a><span class="c1">#SBATCH --account       nesi02659</span>
<a id="__codelineno-1-4" name="__codelineno-1-4"></a><span class="c1">#SBATCH --job-name      variant_calling_workflow</span>
<a id="__codelineno-1-5" name="__codelineno-1-5"></a><span class="c1">#SBATCH --cpus-per-task 2</span>
<a id="__codelineno-1-6" name="__codelineno-1-6"></a><span class="c1">#SBATCH --time          00:15:00</span>
<a id="__codelineno-1-7" name="__codelineno-1-7"></a><span class="c1">#SBATCH --mem           4G</span>
<a id="__codelineno-1-8" name="__codelineno-1-8"></a><span class="c1">#SBATCH --output        slurmout/variant_calling-%j.out</span>
<a id="__codelineno-1-9" name="__codelineno-1-9"></a><span class="c1">#SBATCH --error         slurmout/variant_calling-%j.err</span>
<a id="__codelineno-1-10" name="__codelineno-1-10"></a><span class="c1">#SBATCH --mail-type     END</span>
<a id="__codelineno-1-11" name="__codelineno-1-11"></a><span class="c1">#SBATCH --mail-user     myemail@email.co.nz</span>
<a id="__codelineno-1-12" name="__codelineno-1-12"></a>
<a id="__codelineno-1-13" name="__codelineno-1-13"></a><span class="c1"># Load all the required modules</span>
<a id="__codelineno-1-14" name="__codelineno-1-14"></a>module<span class="w"> </span>purge
<a id="__codelineno-1-15" name="__codelineno-1-15"></a>module<span class="w"> </span>load<span class="w"> </span>BWA/0.7.17-GCC-9.2.0
<a id="__codelineno-1-16" name="__codelineno-1-16"></a>module<span class="w"> </span>load<span class="w"> </span>SAMtools/1.13-GCC-9.2.0
<a id="__codelineno-1-17" name="__codelineno-1-17"></a>module<span class="w"> </span>load<span class="w"> </span>BCFtools/1.13-GCC-9.2.0
<a id="__codelineno-1-18" name="__codelineno-1-18"></a>
<a id="__codelineno-1-19" name="__codelineno-1-19"></a><span class="nb">echo</span><span class="w"> </span><span class="s2">&quot;</span><span class="nv">$PWD</span><span class="s2">&quot;</span>
<a id="__codelineno-1-20" name="__codelineno-1-20"></a>
<a id="__codelineno-1-21" name="__codelineno-1-21"></a><span class="c1"># create the results directories</span>
<a id="__codelineno-1-22" name="__codelineno-1-22"></a>mkdir<span class="w"> </span>-p<span class="w"> </span>results/sam<span class="w"> </span>results/bam<span class="w"> </span>results/bcf<span class="w"> </span>results/vcf
<a id="__codelineno-1-23" name="__codelineno-1-23"></a>
<a id="__codelineno-1-24" name="__codelineno-1-24"></a><span class="c1"># indexing the genome</span>
<a id="__codelineno-1-25" name="__codelineno-1-25"></a><span class="nv">genome</span><span class="o">=</span>~/scripting_workshop/variant_calling/ref_genome/ecoli_rel606.fasta
<a id="__codelineno-1-26" name="__codelineno-1-26"></a><span class="nv">trimmed</span><span class="o">=</span>~/scripting_workshop/variant_calling/trimmed_reads
<a id="__codelineno-1-27" name="__codelineno-1-27"></a>bwa<span class="w"> </span>index<span class="w"> </span><span class="nv">$genome</span>
<a id="__codelineno-1-28" name="__codelineno-1-28"></a>
<a id="__codelineno-1-29" name="__codelineno-1-29"></a><span class="c1"># create a loop that map reads to the genome, sort the bam files and call variants</span>
<a id="__codelineno-1-30" name="__codelineno-1-30"></a><span class="k">for</span><span class="w"> </span>fq1<span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="si">${</span><span class="nv">trimmed</span><span class="si">}</span>/*_1.trim.sub.fastq
<a id="__codelineno-1-31" name="__codelineno-1-31"></a><span class="w"> </span><span class="k">do</span>
<a id="__codelineno-1-32" name="__codelineno-1-32"></a><span class="w">    </span><span class="nb">echo</span><span class="w"> </span><span class="s2">&quot;working with file </span><span class="nv">$fq1</span><span class="s2">&quot;</span>
<a id="__codelineno-1-33" name="__codelineno-1-33"></a>
<a id="__codelineno-1-34" name="__codelineno-1-34"></a><span class="w">    </span><span class="nv">base</span><span class="o">=</span><span class="k">$(</span>basename<span class="w"> </span><span class="nv">$fq1</span><span class="w"> </span>_1.trim.sub.fastq<span class="k">)</span>
<a id="__codelineno-1-35" name="__codelineno-1-35"></a><span class="w">    </span><span class="nb">echo</span><span class="w"> </span><span class="s2">&quot;base name is </span><span class="nv">$base</span><span class="s2">&quot;</span>
<a id="__codelineno-1-36" name="__codelineno-1-36"></a>
<a id="__codelineno-1-37" name="__codelineno-1-37"></a><span class="w">    </span><span class="c1"># setting the variables</span>
<a id="__codelineno-1-38" name="__codelineno-1-38"></a><span class="w">    </span><span class="nv">fq1</span><span class="o">=</span><span class="si">${</span><span class="nv">trimmed</span><span class="si">}</span>/<span class="si">${</span><span class="nv">base</span><span class="si">}</span>_1.trim.sub.fastq
<a id="__codelineno-1-39" name="__codelineno-1-39"></a><span class="w">    </span><span class="nv">fq2</span><span class="o">=</span><span class="si">${</span><span class="nv">trimmed</span><span class="si">}</span>/<span class="si">${</span><span class="nv">base</span><span class="si">}</span>_2.trim.sub.fastq
<a id="__codelineno-1-40" name="__codelineno-1-40"></a><span class="w">    </span><span class="nv">sam</span><span class="o">=</span>results/sam/<span class="si">${</span><span class="nv">base</span><span class="si">}</span>.aligned.sam
<a id="__codelineno-1-41" name="__codelineno-1-41"></a><span class="w">    </span><span class="nv">bam</span><span class="o">=</span>results/bam/<span class="si">${</span><span class="nv">base</span><span class="si">}</span>.aligned.bam
<a id="__codelineno-1-42" name="__codelineno-1-42"></a><span class="w">    </span><span class="nv">sorted_bam</span><span class="o">=</span>results/bam/<span class="si">${</span><span class="nv">base</span><span class="si">}</span>.aligned.sorted.bam
<a id="__codelineno-1-43" name="__codelineno-1-43"></a><span class="w">    </span><span class="nv">raw_bcf</span><span class="o">=</span>results/bcf/<span class="si">${</span><span class="nv">base</span><span class="si">}</span>_raw.bcf
<a id="__codelineno-1-44" name="__codelineno-1-44"></a><span class="w">    </span><span class="nv">variants</span><span class="o">=</span>results/vcf/<span class="si">${</span><span class="nv">base</span><span class="si">}</span>_variants.vcf
<a id="__codelineno-1-45" name="__codelineno-1-45"></a><span class="w">    </span><span class="nv">final_variants</span><span class="o">=</span>results/vcf/<span class="si">${</span><span class="nv">base</span><span class="si">}</span>_final_variants.vcf
<a id="__codelineno-1-46" name="__codelineno-1-46"></a>
<a id="__codelineno-1-47" name="__codelineno-1-47"></a><span class="w">    </span><span class="c1"># running the analysis steps</span>
<a id="__codelineno-1-48" name="__codelineno-1-48"></a><span class="w">    </span>bwa<span class="w"> </span>mem<span class="w"> </span><span class="nv">$genome</span><span class="w"> </span><span class="nv">$fq1</span><span class="w"> </span><span class="nv">$fq2</span><span class="w"> </span>&gt;<span class="w"> </span><span class="nv">$sam</span>
<a id="__codelineno-1-49" name="__codelineno-1-49"></a><span class="w">    </span>samtools<span class="w"> </span>view<span class="w"> </span>-S<span class="w"> </span>-b<span class="w"> </span><span class="nv">$sam</span><span class="w"> </span>&gt;<span class="w"> </span><span class="nv">$bam</span>
<a id="__codelineno-1-50" name="__codelineno-1-50"></a><span class="w">    </span>samtools<span class="w"> </span>sort<span class="w"> </span>-o<span class="w"> </span><span class="nv">$sorted_bam</span><span class="w"> </span><span class="nv">$bam</span>
<a id="__codelineno-1-51" name="__codelineno-1-51"></a><span class="w">    </span>samtools<span class="w"> </span>index<span class="w"> </span><span class="nv">$sorted_bam</span>
<a id="__codelineno-1-52" name="__codelineno-1-52"></a><span class="w">    </span>bcftools<span class="w"> </span>mpileup<span class="w"> </span>-O<span class="w"> </span>b<span class="w"> </span>-o<span class="w"> </span><span class="nv">$raw_bcf</span><span class="w"> </span>-f<span class="w"> </span><span class="nv">$genome</span><span class="w"> </span><span class="nv">$sorted_bam</span>
<a id="__codelineno-1-53" name="__codelineno-1-53"></a><span class="w">    </span>bcftools<span class="w"> </span>call<span class="w"> </span>--ploidy<span class="w"> </span><span class="m">1</span><span class="w"> </span>-m<span class="w"> </span>-v<span class="w"> </span>-o<span class="w"> </span><span class="nv">$variants</span><span class="w"> </span><span class="nv">$raw_bcf</span>
<a id="__codelineno-1-54" name="__codelineno-1-54"></a><span class="w">    </span>vcfutils.pl<span class="w"> </span>varFilter<span class="w"> </span><span class="nv">$variants</span><span class="w"> </span>&gt;<span class="w"> </span><span class="nv">$final_variants</span>
<a id="__codelineno-1-55" name="__codelineno-1-55"></a>
<a id="__codelineno-1-56" name="__codelineno-1-56"></a><span class="k">done</span>
<a id="__codelineno-1-57" name="__codelineno-1-57"></a>
<a id="__codelineno-1-58" name="__codelineno-1-58"></a><span class="nb">echo</span><span class="w"> </span><span class="s2">&quot;DONE&quot;</span>
</code></pre></div></td></tr></table></div>
</details>
<details class="circle-check">
<summary>Exercise 5.5 🐮</summary>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-2-2"> 2</a></span>
<span class="normal"><a href="#__codelineno-2-3"> 3</a></span>
<span class="normal"><a href="#__codelineno-2-4"> 4</a></span>
<span class="normal"><a href="#__codelineno-2-5"> 5</a></span>
<span class="normal"><a href="#__codelineno-2-6"> 6</a></span>
<span class="normal"><a href="#__codelineno-2-7"> 7</a></span>
<span class="normal"><a href="#__codelineno-2-8"> 8</a></span>
<span class="normal"><a href="#__codelineno-2-9"> 9</a></span>
<span class="normal"><a href="#__codelineno-2-10">10</a></span>
<span class="normal"><a href="#__codelineno-2-11">11</a></span>
<span class="normal"><a href="#__codelineno-2-12">12</a></span>
<span class="normal"><a href="#__codelineno-2-13">13</a></span>
<span class="normal"><a href="#__codelineno-2-14">14</a></span>
<span class="normal"><a href="#__codelineno-2-15">15</a></span>
<span class="normal"><a href="#__codelineno-2-16">16</a></span>
<span class="normal"><a href="#__codelineno-2-17">17</a></span>
<span class="normal"><a href="#__codelineno-2-18">18</a></span>
<span class="normal"><a href="#__codelineno-2-19">19</a></span>
<span class="normal"><a href="#__codelineno-2-20">20</a></span>
<span class="normal"><a href="#__codelineno-2-21">21</a></span>
<span class="normal"><a href="#__codelineno-2-22">22</a></span>
<span class="normal"><a href="#__codelineno-2-23">23</a></span>
<span class="normal"><a href="#__codelineno-2-24">24</a></span>
<span class="normal"><a href="#__codelineno-2-25">25</a></span>
<span class="normal"><a href="#__codelineno-2-26">26</a></span>
<span class="normal"><a href="#__codelineno-2-27">27</a></span>
<span class="normal"><a href="#__codelineno-2-28">28</a></span>
<span class="normal"><a href="#__codelineno-2-29">29</a></span>
<span class="normal"><a href="#__codelineno-2-30">30</a></span>
<span class="normal"><a href="#__codelineno-2-31">31</a></span>
<span class="normal"><a href="#__codelineno-2-32">32</a></span>
<span class="normal"><a href="#__codelineno-2-33">33</a></span>
<span class="normal"><a href="#__codelineno-2-34">34</a></span>
<span class="normal"><a href="#__codelineno-2-35">35</a></span>
<span class="normal"><a href="#__codelineno-2-36">36</a></span>
<span class="normal"><a href="#__codelineno-2-37">37</a></span>
<span class="normal"><a href="#__codelineno-2-38">38</a></span>
<span class="normal"><a href="#__codelineno-2-39">39</a></span>
<span class="normal"><a href="#__codelineno-2-40">40</a></span>
<span class="normal"><a href="#__codelineno-2-41">41</a></span>
<span class="normal"><a href="#__codelineno-2-42">42</a></span>
<span class="normal"><a href="#__codelineno-2-43">43</a></span>
<span class="normal"><a href="#__codelineno-2-44">44</a></span>
<span class="normal"><a href="#__codelineno-2-45">45</a></span>
<span class="normal"><a href="#__codelineno-2-46">46</a></span>
<span class="normal"><a href="#__codelineno-2-47">47</a></span>
<span class="normal"><a href="#__codelineno-2-48">48</a></span>
<span class="normal"><a href="#__codelineno-2-49">49</a></span>
<span class="normal"><a href="#__codelineno-2-50">50</a></span>
<span class="normal"><a href="#__codelineno-2-51">51</a></span>
<span class="normal"><a href="#__codelineno-2-52">52</a></span>
<span class="normal"><a href="#__codelineno-2-53">53</a></span>
<span class="normal"><a href="#__codelineno-2-54">54</a></span>
<span class="normal"><a href="#__codelineno-2-55">55</a></span>
<span class="normal"><a href="#__codelineno-2-56">56</a></span>
<span class="normal"><a href="#__codelineno-2-57">57</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-2-2" name="__codelineno-2-2"></a><span class="ch">#!/bin/bash -e</span>
<a id="__codelineno-2-3" name="__codelineno-2-3"></a>
<a id="__codelineno-2-4" name="__codelineno-2-4"></a><span class="c1">#SBATCH --account       nesi02659</span>
<a id="__codelineno-2-5" name="__codelineno-2-5"></a><span class="c1">#SBATCH --job-name      rna-seq_workflow</span>
<a id="__codelineno-2-6" name="__codelineno-2-6"></a><span class="c1">#SBATCH --cpus-per-task 2</span>
<a id="__codelineno-2-7" name="__codelineno-2-7"></a><span class="c1">#SBATCH --time          00:15:00</span>
<a id="__codelineno-2-8" name="__codelineno-2-8"></a><span class="c1">#SBATCH --mem           4G</span>
<a id="__codelineno-2-9" name="__codelineno-2-9"></a><span class="c1">#SBATCH --output        rna-seq_workflow-%j.out</span>
<a id="__codelineno-2-10" name="__codelineno-2-10"></a><span class="c1">#SBATCH --error         rna-seq_workflow-%j.err</span>
<a id="__codelineno-2-11" name="__codelineno-2-11"></a><span class="c1">#SBATCH --mail-type     END</span>
<a id="__codelineno-2-12" name="__codelineno-2-12"></a><span class="c1">#SBATCH --mail-user     myemail@email.org.nz</span>
<a id="__codelineno-2-13" name="__codelineno-2-13"></a>
<a id="__codelineno-2-14" name="__codelineno-2-14"></a>
<a id="__codelineno-2-15" name="__codelineno-2-15"></a><span class="nb">echo</span><span class="w"> </span><span class="s2">&quot;</span><span class="nv">$PWD</span><span class="s2">&quot;</span>
<a id="__codelineno-2-16" name="__codelineno-2-16"></a>
<a id="__codelineno-2-17" name="__codelineno-2-17"></a>mkdir<span class="w"> </span>-p<span class="w"> </span>~/scripting_workshop/scheduler/ex_5.5/<span class="o">{</span>Mapping,Counts<span class="o">}</span><span class="w"> </span><span class="o">&amp;&amp;</span><span class="w"> </span><span class="nb">cd</span><span class="w"> </span>~/scripting_workshop/scheduler/ex_5.5/
<a id="__codelineno-2-18" name="__codelineno-2-18"></a>cp<span class="w"> </span>-r<span class="w"> </span>/nesi/project/nesi02659/scripting_workshop/rna_seq/*<span class="w"> </span>./
<a id="__codelineno-2-19" name="__codelineno-2-19"></a>
<a id="__codelineno-2-20" name="__codelineno-2-20"></a>module<span class="w"> </span>purge
<a id="__codelineno-2-21" name="__codelineno-2-21"></a>module<span class="w"> </span>load<span class="w"> </span>HISAT2/2.2.0-gimkl-2020a
<a id="__codelineno-2-22" name="__codelineno-2-22"></a>module<span class="w"> </span>load<span class="w"> </span>SAMtools/1.10-GCC-9.2.0
<a id="__codelineno-2-23" name="__codelineno-2-23"></a>module<span class="w"> </span>load<span class="w"> </span>Subread/2.0.0-GCC-9.2.0
<a id="__codelineno-2-24" name="__codelineno-2-24"></a>
<a id="__codelineno-2-25" name="__codelineno-2-25"></a><span class="nb">echo</span><span class="w"> </span><span class="nv">$PWD</span>
<a id="__codelineno-2-26" name="__codelineno-2-26"></a>
<a id="__codelineno-2-27" name="__codelineno-2-27"></a><span class="c1">#index file</span>
<a id="__codelineno-2-28" name="__codelineno-2-28"></a>hisat2-build<span class="w"> </span>-p<span class="w"> </span><span class="m">4</span><span class="w"> </span>-f<span class="w"> </span><span class="nv">$PWD</span>/ref_genome/Saccharomyces_cerevisiae.R64-1-1.dna.toplevel.fa<span class="w"> </span><span class="nv">$PWD</span>/ref_genome/Saccharomyces_cerevisiae.R64-1-1.dna.toplevel
<a id="__codelineno-2-29" name="__codelineno-2-29"></a>
<a id="__codelineno-2-30" name="__codelineno-2-30"></a>
<a id="__codelineno-2-31" name="__codelineno-2-31"></a><span class="c1">#Mapping Samples to the reference genome</span>
<a id="__codelineno-2-32" name="__codelineno-2-32"></a>
<a id="__codelineno-2-33" name="__codelineno-2-33"></a><span class="k">for</span><span class="w"> </span>filename<span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="nv">$PWD</span>/trimmed_reads/*
<a id="__codelineno-2-34" name="__codelineno-2-34"></a><span class="w">  </span><span class="k">do</span>
<a id="__codelineno-2-35" name="__codelineno-2-35"></a><span class="w">    </span><span class="nv">base</span><span class="o">=</span><span class="k">$(</span>basename<span class="w"> </span><span class="si">${</span><span class="nv">filename</span><span class="si">}</span><span class="w"> </span>.fastq<span class="k">)</span>
<a id="__codelineno-2-36" name="__codelineno-2-36"></a><span class="w">    </span>hisat2<span class="w"> </span>-p<span class="w"> </span><span class="m">4</span><span class="w"> </span>-x<span class="w"> </span><span class="nv">$PWD</span>/ref_genome/Saccharomyces_cerevisiae.R64-1-1.dna.toplevel<span class="w"> </span>-U<span class="w"> </span><span class="nv">$filename</span><span class="w"> </span>-S<span class="w"> </span><span class="nv">$PWD</span>/Mapping/<span class="si">${</span><span class="nv">base</span><span class="si">}</span>.sam<span class="w"> </span>--summary-file<span class="w"> </span><span class="nv">$PWD</span>/Mapping/<span class="si">${</span><span class="nv">base</span><span class="si">}</span>_summary.txt
<a id="__codelineno-2-37" name="__codelineno-2-37"></a><span class="w">  </span><span class="k">done</span>
<a id="__codelineno-2-38" name="__codelineno-2-38"></a>
<a id="__codelineno-2-39" name="__codelineno-2-39"></a><span class="c1">#Convert SAMfiles to BAM</span>
<a id="__codelineno-2-40" name="__codelineno-2-40"></a>
<a id="__codelineno-2-41" name="__codelineno-2-41"></a><span class="k">for</span><span class="w"> </span>filename<span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="nv">$PWD</span>/Mapping/*.sam
<a id="__codelineno-2-42" name="__codelineno-2-42"></a><span class="w">  </span><span class="k">do</span>
<a id="__codelineno-2-43" name="__codelineno-2-43"></a><span class="w">   </span><span class="nv">base</span><span class="o">=</span><span class="k">$(</span>basename<span class="w"> </span><span class="si">${</span><span class="nv">filename</span><span class="si">}</span><span class="w"> </span>.sam<span class="k">)</span>
<a id="__codelineno-2-44" name="__codelineno-2-44"></a><span class="w">   </span>samtools<span class="w"> </span>view<span class="w"> </span>-S<span class="w"> </span>-b<span class="w"> </span><span class="si">${</span><span class="nv">filename</span><span class="si">}</span><span class="w"> </span>-o<span class="w"> </span><span class="nv">$PWD</span>/Mapping/<span class="si">${</span><span class="nv">base</span><span class="si">}</span>.bam
<a id="__codelineno-2-45" name="__codelineno-2-45"></a><span class="w">  </span><span class="k">done</span>
<a id="__codelineno-2-46" name="__codelineno-2-46"></a>
<a id="__codelineno-2-47" name="__codelineno-2-47"></a><span class="c1">#Sort BAM files</span>
<a id="__codelineno-2-48" name="__codelineno-2-48"></a>
<a id="__codelineno-2-49" name="__codelineno-2-49"></a><span class="k">for</span><span class="w"> </span>filename<span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="nv">$PWD</span>/Mapping/*.bam
<a id="__codelineno-2-50" name="__codelineno-2-50"></a><span class="w">  </span><span class="k">do</span>
<a id="__codelineno-2-51" name="__codelineno-2-51"></a><span class="w">    </span><span class="nv">base</span><span class="o">=</span><span class="k">$(</span>basename<span class="w"> </span><span class="si">${</span><span class="nv">filename</span><span class="si">}</span><span class="w"> </span>.bam<span class="k">)</span>
<a id="__codelineno-2-52" name="__codelineno-2-52"></a><span class="w">    </span>samtools<span class="w"> </span>sort<span class="w"> </span>-o<span class="w"> </span><span class="nv">$PWD</span>/Mapping/<span class="si">${</span><span class="nv">base</span><span class="si">}</span>_sorted.bam<span class="w"> </span><span class="si">${</span><span class="nv">filename</span><span class="si">}</span>
<a id="__codelineno-2-53" name="__codelineno-2-53"></a><span class="w">  </span><span class="k">done</span>
<a id="__codelineno-2-54" name="__codelineno-2-54"></a>
<a id="__codelineno-2-55" name="__codelineno-2-55"></a><span class="c1">#count how many reads aligned to each genome feature (exon).</span>
<a id="__codelineno-2-56" name="__codelineno-2-56"></a>
<a id="__codelineno-2-57" name="__codelineno-2-57"></a>featureCounts<span class="w"> </span>-a<span class="w"> </span><span class="nv">$PWD</span>/ref_genome/Saccharomyces_cerevisiae.R64-1-1.99.gtf<span class="w"> </span>-o<span class="w"> </span><span class="nv">$PWD</span>/Counts/yeast_counts.txt<span class="w"> </span>-T<span class="w"> </span><span class="m">2</span><span class="w"> </span>-t<span class="w"> </span>exon<span class="w"> </span>-g<span class="w"> </span>gene_id<span class="w"> </span><span class="nv">$PWD</span>/Mapping/*sorted.bam
</code></pre></div></td></tr></table></div>
</details>
<p align="center"><b><a href="https://genomicsaotearoa.github.io/Workshop-Bash_Scripting_And_HPC_Job_Scheduler/">Back to homepage</a></b></p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer" >
        
          
          <a href="../7_supplementary_2/" class="md-footer__link md-footer__link--prev" aria-label="Previous: S2 : slurm profiling">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                S2 : slurm profiling
              </div>
            </div>
          </a>
        
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Introduction to Bash Scripting and HPC Scheduler is licensed under a <a rel="noopener" target="_blank" href="https://opensource.org/license/mit/">MIT License</a>
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "..", "features": ["navigation.tracking", "navigation.sections", "navigation.expand", "navigation.top", "navigation.footer", "content.code.copy", "content.code.annotate"], "search": "../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.471ce7a9.min.js"></script>
      
        <script src="../javascripts/mathjax.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  <script id="init-glightbox">const lightbox = GLightbox({"touchNavigation": true, "loop": false, "zoomable": true, "draggable": true, "openEffect": "zoom", "closeEffect": "zoom", "slideEffect": "slide"});
document$.subscribe(() => { lightbox.reload() });
</script></body>
</html>